# Ensemble Learning

## Initial Imports

In [1]:
import warnings
warnings.filterwarnings('ignore')

In [2]:
import numpy as np
import pandas as pd
from pathlib import Path
from collections import Counter

In [48]:
from sklearn.metrics import balanced_accuracy_score
from sklearn.metrics import confusion_matrix
from imblearn.metrics import classification_report_imbalanced

from sklearn.metrics import confusion_matrix, accuracy_score, classification_report

## Read the CSV and Perform Basic Data Cleaning

In [49]:
# Load the data
file_path = Path('Resources/LoanStats_2019Q1.csv')
df = pd.read_csv(file_path)

# Preview the data
df.head()

loan_amnt  int_rate  installment home_ownership  annual_inc  \
0    10500.0    0.1719       375.35           RENT     66000.0   
1    25000.0    0.2000       929.09       MORTGAGE    105000.0   
2    20000.0    0.2000       529.88       MORTGAGE     56000.0   
3    10000.0    0.1640       353.55           RENT     92000.0   
4    22000.0    0.1474       520.39       MORTGAGE     52000.0   

  verification_status   issue_d loan_status pymnt_plan    dti  ...  \
0     Source Verified  Mar-2019    low_risk          n  27.24  ...   
1            Verified  Mar-2019    low_risk          n  20.23  ...   
2            Verified  Mar-2019    low_risk          n  24.26  ...   
3            Verified  Mar-2019    low_risk          n  31.44  ...   
4        Not Verified  Mar-2019    low_risk          n  18.76  ...   

   pct_tl_nvr_dlq  percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  \
0            85.7             100.0                   0.0        0.0   
1            91.2              50.0                   1.0        0.0   
2            66.7              50.0                   0.0        0.0   
3           100.0              50.0                   1.0        0.0   
4           100.0               0.0                   0.0        0.0   

   tot_hi_cred_lim  total_bal_ex_mort total_bc_limit  \
0          65687.0            38199.0         2000.0   
1         271427.0            60641.0        41200.0   
2          60644.0            45684.0         7500.0   
3          99506.0            68784.0        19700.0   
4         219750.0            25919.0        27600.0   

   total_il_high_credit_limit  hardship_flag  debt_settlement_flag  
0                     61987.0              N                     N  
1                     49197.0              N                     N  
2                     43144.0              N                     N  
3                     76506.0              N                     N  
4                     20000.0              N                     N  

[5 rows x 86 columns]

## Split the Data into Training and Testing

In [50]:
# Create our features
X = df.copy()
X.drop("loan_status", axis = 1, inplace=True)
X.head()

# Create our target
y = df["loan_status"].values.reshape(-1, 1)




In [51]:
y = df["loan_status"]

# Create our features
X = df.drop(columns="loan_status")
X = X.select_dtypes(['number'])
X.head()



loan_amnt  int_rate  installment  annual_inc    dti  delinq_2yrs  \
0    10500.0    0.1719       375.35     66000.0  27.24          0.0   
1    25000.0    0.2000       929.09    105000.0  20.23          0.0   
2    20000.0    0.2000       529.88     56000.0  24.26          0.0   
3    10000.0    0.1640       353.55     92000.0  31.44          0.0   
4    22000.0    0.1474       520.39     52000.0  18.76          0.0   

   inq_last_6mths  open_acc  pub_rec  revol_bal  ...  num_tl_90g_dpd_24m  \
0             0.0       8.0      0.0     1609.0  ...                 0.0   
1             0.0      17.0      1.0    18368.0  ...                 0.0   
2             0.0       8.0      0.0    13247.0  ...                 0.0   
3             1.0      10.0      1.0    17996.0  ...                 0.0   
4             1.0      14.0      0.0     9091.0  ...                 0.0   

   num_tl_op_past_12m  pct_tl_nvr_dlq  percent_bc_gt_75  pub_rec_bankruptcies  \
0                 3.0            85.7             100.0                   0.0   
1                 6.0            91.2              50.0                   1.0   
2                 4.0            66.7              50.0                   0.0   
3                 1.0           100.0              50.0                   1.0   
4                 1.0           100.0               0.0                   0.0   

   tax_liens  tot_hi_cred_lim  total_bal_ex_mort  total_bc_limit  \
0        0.0          65687.0            38199.0          2000.0   
1        0.0         271427.0            60641.0         41200.0   
2        0.0          60644.0            45684.0          7500.0   
3        0.0          99506.0            68784.0         19700.0   
4        0.0         219750.0            25919.0         27600.0   

   total_il_high_credit_limit  
0                     61987.0  
1                     49197.0  
2                     43144.0  
3                     76506.0  
4                     20000.0  

[5 rows x 76 columns]

In [52]:
X.describe()

loan_amnt      int_rate   installment    annual_inc           dti  \
count  68817.000000  68817.000000  68817.000000  6.881700e+04  68817.000000   
mean   16677.594562      0.127718    480.652863  8.821371e+04     21.778153   
std    10277.348590      0.048130    288.062432  1.155800e+05     20.199244   
min     1000.000000      0.060000     30.890000  4.000000e+01      0.000000   
25%     9000.000000      0.088100    265.730000  5.000000e+04     13.890000   
50%    15000.000000      0.118000    404.560000  7.300000e+04     19.760000   
75%    24000.000000      0.155700    648.100000  1.040000e+05     26.660000   
max    40000.000000      0.308400   1676.230000  8.797500e+06    999.000000   

        delinq_2yrs  inq_last_6mths      open_acc       pub_rec  \
count  68817.000000    68817.000000  68817.000000  68817.000000   
mean       0.217766        0.497697     12.587340      0.126030   
std        0.718367        0.758122      6.022869      0.336797   
min        0.000000        0.000000      2.000000      0.000000   
25%        0.000000        0.000000      8.000000      0.000000   
50%        0.000000        0.000000     11.000000      0.000000   
75%        0.000000        1.000000     16.000000      0.000000   
max       18.000000        5.000000     72.000000      4.000000   

           revol_bal  ...  num_tl_90g_dpd_24m  num_tl_op_past_12m  \
count   68817.000000  ...        68817.000000        68817.000000   
mean    17604.142828  ...            0.052138            2.219423   
std     21835.880400  ...            0.390633            1.897432   
min         0.000000  ...            0.000000            0.000000   
25%      6293.000000  ...            0.000000            1.000000   
50%     12068.000000  ...            0.000000            2.000000   
75%     21735.000000  ...            0.000000            3.000000   
max    587191.000000  ...           18.000000           19.000000   

       pct_tl_nvr_dlq  percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  \
count    68817.000000      68817.000000          68817.000000    68817.0   
mean        95.057627         30.626217              0.125972        0.0   
std          8.326426         33.631463              0.336732        0.0   
min         20.000000          0.000000              0.000000        0.0   
25%         93.000000          0.000000              0.000000        0.0   
50%        100.000000         20.000000              0.000000        0.0   
75%        100.000000         50.000000              0.000000        0.0   
max        100.000000        100.000000              4.000000        0.0   

       tot_hi_cred_lim  total_bal_ex_mort  total_bc_limit  \
count     6.881700e+04       6.881700e+04    68817.000000   
mean      2.100332e+05       6.133843e+04    29734.128558   
std       1.928088e+05       5.738798e+04    26795.394232   
min       3.600000e+03       2.350000e+02      100.000000   
25%       6.697700e+04       2.650300e+04    11600.000000   
50%       1.467100e+05       4.535700e+04    22100.000000   
75%       3.036400e+05       7.657000e+04    39300.000000   
max       3.292782e+06       1.295455e+06   509400.000000   

       total_il_high_credit_limit  
count                6.881700e+04  
mean                 5.572240e+04  
std                  5.095845e+04  
min                  1.270000e+02  
25%                  2.288000e+04  
50%                  4.200000e+04  
75%                  7.249900e+04  
max                  1.426964e+06  

[8 rows x 76 columns]

In [53]:
# Check the balance of our target values
# YOUR CODE HERE
y[60000:60010]



60000     low_risk
60001     low_risk
60002     low_risk
60003     low_risk
60004    high_risk
60005     low_risk
60006     low_risk
60007     low_risk
60008     low_risk
60009     low_risk
Name: loan_status, dtype: object

In [54]:
# Split the X and y into X_train, X_test, y_train, y_test
# YOUR CODE HERE
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

## Data Pre-Processing

Scale the training and testing data using the `StandardScaler` from `sklearn`. Remember that when scaling the data, you only scale the features data (`X_train` and `X_testing`).

In [55]:
# Create the StandardScaler instance
# YOUR CODE HERE
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()

In [56]:
# Fit the Standard Scaler with the training data
# When fitting scaling functions, only train on the training dataset
# YOUR CODE HERE

X_scaler = scaler.fit(X_train)

In [57]:
# Scale the training and testing data
# YOUR CODE HERE

X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

## Ensemble Learners

In this section, you will compare two ensemble algorithms to determine which algorithm results in the best performance. You will train a Balanced Random Forest Classifier and an Easy Ensemble classifier . For each algorithm, be sure to complete the folliowing steps:

1. Train the model using the training data. 
2. Calculate the balanced accuracy score from sklearn.metrics.
3. Display the confusion matrix from sklearn.metrics.
4. Generate a classication report using the `imbalanced_classification_report` from imbalanced-learn.
5. For the Balanced Random Forest Classifier only, print the feature importance sorted in descending order (most important feature to least important) along with the feature score

Note: Use a random state of 1 for each algorithm to ensure consistency between tests

### Balanced Random Forest Classifier

In [58]:
# Resample the training data with the BalancedRandomForestClassifier
# YOUR CODE HERE

from sklearn.ensemble import RandomForestClassifier
scaler = StandardScaler()
X_scaler = scaler.fit(X_train)
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

# as per instructions 
rf_model = RandomForestClassifier(n_estimators=100, random_state=1)
rf_model = rf_model.fit(X_train_scaled, y_train)


In [59]:
predictions = rf_model.predict(X_test_scaled)



In [60]:
# Calculated the balanced accuracy score
# YOUR CODE HERE

accu_score = accuracy_score(y_test, predictions)

print(f"Accuracy Score : {accu_score}")


Accuracy Score : 0.9955826794536472


In [61]:
# Display the confusion matrix
# YOUR CODE HERE
cm = confusion_matrix(y_test, predictions)
cm_df = pd.DataFrame(
    cm, index=["Actual 0", "Actual 1"], columns=["Predicted 0", "Predicted 1"]
)

# Displaying results
print("Confusion Matrix")
display(cm_df)


Confusion Matrix


Predicted 0  Predicted 1
Actual 0           32           69
Actual 1            7        17097

In [64]:
# Print the imbalanced classification report
# YOUR CODE HERE

print("Classification Report")
print(classification_report(y_test, predictions))

Classification Report
              precision    recall  f1-score   support

   high_risk       0.82      0.32      0.46       101
    low_risk       1.00      1.00      1.00     17104

    accuracy                           1.00     17205
   macro avg       0.91      0.66      0.73     17205
weighted avg       0.99      1.00      0.99     17205



In [65]:
# List the features sorted in descending order by feature importance
# YOUR CODE HERE
importances = rf_model.feature_importances_
importances_sorted = sorted(zip(rf_model.feature_importances_, X.columns), reverse=True)
importances_sorted[:10]




[(0.06972106761227882, 'total_rec_int'),
 (0.06694225660766305, 'total_pymnt_inv'),
 (0.06326625735854315, 'last_pymnt_amnt'),
 (0.05949173381269891, 'total_rec_prncp'),
 (0.04727264125614207, 'total_pymnt'),
 (0.019776309380222564, 'dti'),
 (0.01909962919701392, 'max_bal_bc'),
 (0.018837666058058252, 'mo_sin_old_rev_tl_op'),
 (0.018402807461673262, 'installment'),
 (0.018360105060012125, 'bc_open_to_buy')]

### Easy Ensemble Classifier

In [66]:
# Train the Classifier
# YOUR CODE HERE
from imblearn.ensemble import EasyEnsembleClassifier 

scaler = StandardScaler()
X_scaler = scaler.fit(X_train)
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

# as per instructions 
rf_model =EasyEnsembleClassifier(n_estimators=100, random_state=1)
rf_model = rf_model.fit(X_train_scaled, y_train)


In [67]:
# Calculated the balanced accuracy score
# YOUR CODE HERE

predictions = rf_model.predict(X_test_scaled)

# Calculated the balanced accuracy score
acc_score = accuracy_score(y_test, predictions)
acc_score

0.7634408602150538

In [68]:
# Display the confusion matrix
# YOUR CODE HERE
cm = confusion_matrix(y_test, predictions)
cm_df = pd.DataFrame(
    cm, index=["Actual 0", "Actual 1"], columns=["Predicted 0", "Predicted 1"]
)

# Calculating the accuracy score
acc_score = accuracy_score(y_test, predictions)
acc_score


# Displaying results
print("Confusion Matrix")
display(cm_df)
print(f"Accuracy Score : {acc_score}")




Confusion Matrix


Predicted 0  Predicted 1
Actual 0           70           31
Actual 1         4039        13065

Accuracy Score : 0.7634408602150538


In [69]:
# Print the imbalanced classification report
# YOUR CODE HERE

print("Classification Report")
print(classification_report(y_test, predictions))

Classification Report
              precision    recall  f1-score   support

   high_risk       0.02      0.69      0.03       101
    low_risk       1.00      0.76      0.87     17104

    accuracy                           0.76     17205
   macro avg       0.51      0.73      0.45     17205
weighted avg       0.99      0.76      0.86     17205



### Final Questions
1. Which model had the best balanced accuracy score?

    Balanced Random Forest Classifier 0.99500

2. Which model had the best recall score?

    Both Balanced Random Forest Classifier and Easy Ensemble Classifier are the same with a macro average of 0.66

3. Which model had the best geometric mean score?
 
    YOUR ANSWER HERE.

4. What are the top three features?

  (0.08097838943943801, 'total_pymnt_inv'),
 (0.07812562105474924, 'total_rec_int'),
 (0.07482443322056208, 'last_pymnt_amnt'),